### ISSUE TITLE
Scoreboards Auto Complete regardless, also a game breaking crash when resetting quests.

### ISSUE THREAD

**Issue by shade28** [Funwayguy/BetterQuesting#374]
When developing my modpack, i stumbled upon several issues with better questing. I really like that i can run minecraft commands and change scoreboards and such, to me it makes it the Superior quest mod out there. However, when you setup a task requiring a scoreboard several things happen that are not desirable.

If you have the isSilent set to true, the quest will still do the quest update display/level up sounds and the quest will autocomplete even if you didnt want it to. This means that you couldnt use a scoreboard for any kind of currency system.  If say the quest was suppose to reduce the scoreboard as if you spent it. so that you needed to save up for other quests again, instead what happens is ALL quests you can complete that use a scoreboard task do. Spamming your screen with quest updates if you have many. To make this even worse, if you set the quest to auto complete, it will send an update EVERY time the quest refreshes. this can easily result in an infinite loop of quest updates that persist even if you delete the quest. the only way to remove the update messages is to uninstall the mod at that point.

So to try to implement my quest point currency system I decided to try to think outside the box. Using vanilla function files i would set a scoreboard as a flag to update the quests, and then a function would use the command "/bq_admin reset" using a long list of condition checks. this worked fairly well, even if it was a work around. however it turns out if you try to reset multiple quests in the same tick you get a client crash as soon as you go to the main page of the quest book. This crash persists each and every time you go to the quest book after reloading your client too. The fix is to "bq_admin default load" reloading all quests as if the quest data became corrupted. I used timers to make sure i only send 1 reset at a time, however some players of my pack still get this bug from time to time. 

I am hoping that you could fix these annoyances in an otherwise amazing mod. I Love better questing and hope you continue to improve it.

shade28
https://minecraft.curseforge.com/projects/skytouched

**Comment by Funwayguy** [comment]
I believe I only just fixed the incorrect update notifications. As for using a scoreboard task as a currency system, might I recommend having a checkbox and scoreboard task in simultaneous mode and then using a scoreboard reward to do the subtraction. Setting the repeat time to about 10 ticks should let it reset itself every half second (when necessary).

I am aware of problems with these kinds of quest setups but due to fundamental issues with the old code in Better Questing it will take some time to rip it all up and rewrite it. I'm currently only focusing on the visual side of things so they're robust enough to take the core changes when they're rolled out.

**Comment by shade28** [comment]
The issue with scoreboards plus checkboxes, is that the scoreboards still count as complete on all quests that use that score even if your score drops below the threshold in the future. so adding checkboxes doesnt actually fix the underlying issue. Im just submitting the report, not expecting a quick fix.

**Comment by Funwayguy** [comment]
Just about to update this with that. Was testing and yea there is a lot of broken things with it. Not particularly sure why it's like that. Could be a combination of issues between the base mod not resetting things properly and the expansion incorrectly syncing score data. I might look into this more later depending on how serious the issue is under the surface but it's not a high priority right now.

**Comment by Funwayguy** [comment]
Made some notable fixes to quest updating in build 255 which I've just uploaded to Curse. Can you try again with that build and see how that works out? From what I've tested myself scoreboards appears to work as intended again but I wanted your perspective to make sure.

NOTE: The crash with quests force resetting themselves is probably not going to change nor is it something I intend to support just yet.

**Comment by shade28** [comment]
In my testing it looks good so far. I am going to need to test it for awhile, but ill update here if i find anything.

**Comment by shade28** [comment]
Scoreboards are working so much better, however here is my findings after testing.

I setup a quest to require turning in items, and a scoreboard. the quest reward gives items and removes the same amount of score as the quest required. There are 90 quests that do this.

When you reach a score threshold you can then turn in a quest and your score will be reduced by that amount (lets say you had 3, and the quest required 2, so you now have 1) now if you go to another quest that required 2, since you originally had 3, you can still complete the quest lowering your score to -1.

So in conclusion, scoreboards now reflect the proper score, they honor the isSilent flag, but they still turn themselves in funky.

Thank you again for your hard work.

**Comment by Funwayguy** [comment]
When I tested scoreboards as currency I specifically setup any point deduction quests as autoclaim so users couldn't confirm multiple purchases before the deductions take place. The simultaneous mode also prevents the task from staying completed if the value drops below the threshold again.

NOTE: There is a known issue with the checkbox task not resetting itself properly. This means someone can accidentally lock in their purchase attempt and as soon as the threshold is reached, the transaction will be be automatically completed with no way to cancel out of it.

**Comment by shade28** [comment]
I didnt know about the simultaneous mode under properties when i did my test. Thank you for letting me know. From what I can tell that fixed the issue i was having. when using Items however, if you turn in the quest, and dont have the score it will take and delete the items without giving you credit, Im much happier with this then the alternative but it may be something to address in the future.

**Comment by Funwayguy** [comment]
It might be more worth while if I setup an entire expansion dedicated to commerce and trading rather than rigging quests try and do the same. Scoreboards and quests are better at tracking data, not so much exchanging resources.

Marking this as closed now that the main issue is resolved (not fixing that crash though)
